Mathlib/Algebra/Polynomial/Factors.lean

```diff
@@ -0,0 +1,284 @@
+/-
+Copyright (c) 2025 Thomas Browning. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Thomas Browning
+-/
+import Mathlib.Algebra.Polynomial.FieldDivision
+
+/-!
+# Split polynomials
+
+A polynomial `f : R[X]` factors if it is a product of constant and monic linear polynomials.
+
+## Main definitions
+
+* `Polynomial.Factors f`: A predicate on a polynomial `f` saying that `f` is a product of
+  constant and monic linear polynomials.
+
+-/
+
+variable {R : Type*}
+
+namespace Polynomial
+
+section Semiring
+
+variable [Semiring R]
+
+/-- A polynomial `Factors` if it is a product of constant and monic linear polynomials.
+This will eventually replace `Polynomial.Splits`. -/
+def Factors (f : R[X]) : Prop := f ∈ Submonoid.closure ({C a | a : R} ∪ {X + C a | a : R})
+
+@[simp, aesop safe apply]
+theorem factors_C (a : R) : Factors (C a) :=
+  Submonoid.mem_closure_of_mem (Set.mem_union_left _ ⟨a, rfl⟩)
+
+@[simp, aesop safe apply]
+theorem factors_zero : Factors (0 : R[X]) := by
+  simpa using factors_C (0 : R)
+
+@[simp, aesop safe apply]
+theorem factors_one : Factors (1 : R[X]) :=
+  factors_C (1 : R)
+
+@[simp, aesop safe apply]
+theorem factors_X_add_C (a : R) : Factors (X + C a) :=
+  Submonoid.mem_closure_of_mem (Set.mem_union_right _ ⟨a, rfl⟩)
+
+@[simp, aesop safe apply]
+theorem factors_X : Factors (X : R[X]) := by
+  simpa using factors_X_add_C (0 : R)
+
+@[simp, aesop safe apply]
+protected theorem Factors.mul {f g : R[X]} (hf : Factors f) (hg : Factors g) :
+    Factors (f * g) :=
+  mul_mem hf hg
+
+theorem Factors.C_mul {f : R[X]} (hf : Factors f) (a : R) : Factors (C a * f) :=
+  (factors_C a).mul hf
+
+@[simp, aesop safe apply]
+theorem Factors.list_prod {l : List R[X]} (h : ∀ f ∈ l, Factors f) : Factors l.prod :=
+  list_prod_mem h
+
+@[simp, aesop safe apply]
+protected theorem Factors.pow {f : R[X]} (hf : Factors f) (n : ℕ) : Factors (f ^ n) :=
+  pow_mem hf n
+
+theorem factors_X_pow (n : ℕ) : Factors (X ^ n : R[X]) :=
+  factors_X.pow n
+
+theorem factors_C_mul_X_pow (a : R) (n : ℕ) : Factors (C a * X ^ n) :=
+  (factors_X_pow n).C_mul a
+
+@[simp, aesop safe apply]
+theorem factors_monomial (n : ℕ) (a : R) : Factors (monomial n a) := by
+  simp [← C_mul_X_pow_eq_monomial]
+
+protected theorem Factors.map {f : R[X]} (hf : Factors f) {S : Type*} [Semiring S] (i : R →+* S) :
+    Factors (map i f) := by
+  induction hf using Submonoid.closure_induction <;> aesop
+
+theorem factors_of_natDegree_eq_zero {f : R[X]} (hf : natDegree f = 0) :
+    Factors f :=
+  (natDegree_eq_zero.mp hf).choose_spec ▸ by aesop
+
+theorem factors_of_isUnit [NoZeroDivisors R] {f : R[X]} (hf : IsUnit f) : Factors f :=
+  factors_of_natDegree_eq_zero (natDegree_eq_zero_of_isUnit hf)
```

Comment:
This lemma looks quite weird to me. This is just a simple combination of two lemmas. I don't think we would see `IsUnit f` for `f` a polynomial that often, and I don't think we should encourage the use of `IsUnit f` by adding this lemma. Unless this is somehow a very common application?